### DISCUSSION TITLE
Combining state value and actions

### DISCUSSION THREAD

**Discussion by ZennerBlade** [pmndrs/zustand#1449] [Q&A] [answered]
Hi,
I had a confusion regarding how we should use multiple selectors in-case of selecting state value and action together.

Take this example

```JS
const useBearStore = create((set) => ({
  bears: 0,
  fish: 0,
  increasePopulation: () => set((state) => ({ bears: state.bears + 1 })),
  removeAllBears: () => set({ bears: 0 }),
}))
```
If i want to use _bears, fish_ and _increasePopulation_ together, will this work efficiently

```JS
const { bears, fish , increasePopulation } = useBearStore((state) => ({
bears: state.bears,
fish: state.fish,
increasePopulation: state.increasePopulation,
}), shallow)
```
or is it pointless since a function will always result in equality check failing (as seems from shallow.ts file)

So should i use it this way or separate out states and actions like this
```JS
const { bears, fish , increasePopulation } = useBearStore((state) => ({
bears: state.bears,
fish: state.fish
}), shallow)

const { increasePopulation } = useBearStore((state) => state.increasePopulation)
```

Seems I'll be writing multiple hooks just to extract single action

**Comment by dai-shi** [comment, marked as answer]
```js
const { increasePopulation } = useBearStore((state) => state.increasePopulation)
```
is fine as a pattern, but if you think it's pointless, we can do this:
```js
const { increasePopulation } = useBearStore.getState()
```

Or, you might like this: https://github.com/pmndrs/zustand/blob/2b29d736841dc7b3fd7dec8cbfea50fee7295974/docs/guides/practice-with-no-store-actions.md

**Reply by ZennerBlade** [reply]
So is the assumption correct that combining state value and action together isn't good?

Like this would be a wrong way of using it since my component would rerender when calling _increasePopulation_ even though there's no watch for _bear_ as the object will be created new every render?
```JS
const { fish , increasePopulation } = useBearStore((state) => ({
  fish: state.fish,
  increasePopulation: state.increasePopulation,
}), shallow )
```

Because if that's the case then i can create separate action for hooks or do it like it's mentioned here
https://tkdodo.eu/blog/working-with-zustand#separate-actions-from-state
Basically separating actions in a different object and destructuring from there.

**Reply by dai-shi** [reply]
I think I misunderstood your point of discussion.

> So is the assumption correct that combining state value and action together isn't good?

You mean using `shallow`? It's also good.

```js
const { bears, fish , increasePopulation } = useBearStore((state) => ({
  bears: state.bears,
  fish: state.fish,
  increasePopulation: state.increasePopulation,
}), shallow)
```

So, that ☝️ is good.

> ```js
> const { bears, fish , increasePopulation } = useBearStore((state) => ({
>   fish: state.fish,
>   increasePopulation: state.increasePopulation,
> }), shallow )
> ```

That ☝️ seems bad because `bears` will be `undefined` always.

**Reply by ZennerBlade** [reply]
Sorry the bears in 2nd part was a typo, i meant to delete that as well.

But I get your point. 

My only confusion was that since zustand checks for strict equality, adding shallow wouldn't work since _increasePopulation_ would result in objects not being equal even if _bears_ and _fish_ didn't change

Wouldn't this make the shallow check always fail?
```JS
oldObj.increasePopulation === newObj.increasePopulation   X
```

**Reply by dai-shi** [reply]
> Wouldn't this make the shallow check always fail?
> 
> ```js
> oldObj.increasePopulation === newObj.increasePopulation   X
> ```

No, no. That should usually be `true`.

**Reply by ZennerBlade** [reply]
Thank you so much!!

**Reply by ZennerBlade** [reply]
I had gone by [shallow.ts](https://github.com/pmndrs/zustand/blob/main/src/shallow.ts) and i tried calling it this way
```JS
shallow( {a:()=>{}}, {a:()=>{}} )
```
which gave false, saying objects weren't equal, but this was expected. 
Since _increasePopulation_ is a function of the same type as my test above, i expected it'd give the same result.

I'm guessing Zustand handles it differently internally, right?

**Reply by dai-shi** [reply]
Try:
```js
const fn = () => {}
shallow({ a: fn }, { a: fn })
```
